### PR TITLE
Issue #19: Integrate catalog changes, also fixes issue #17 (multiple images)

### PR DESCRIPTION
**Pull request by Vinai**

### CODE REVIEW

**Comment by Vinai** [comment]
The tests should run once https://github.com/lizards-and-pumpkins/catalog/pull/458 is merged into catalog.

File: src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php [hunk 1]
```diff
@@ -2,34 +2,65 @@
...
+     */
+    private function generateNumberOfCategoriesForProduct($numberOfCategories)
+    {
+        return array_filter([0, 1, 1, 2, 3], function ($nCategories) use ($numberOfCategories) {
+            return $nCategories <= $numberOfCategories;
+        });
+    }
+
+    /**
+     * @return string]|
```
Comment (tim-bezhashvyly) [inline src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php]: A typo.

File: src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php [hunk 2]
```diff
@@ -100,17 +129,17 @@ public function getNumberOfCategories()
     }
 
     /**
-     * @param string[] $contextDataSet
+     * @param Product[] $productInOtherContexts
      * @param string[] $otherValues
      * @return string
      */
-    public function getCategory(array $contextDataSet, array $otherValues)
+    public function getCategory(array $productInOtherContexts, array $otherValues)
```
Comment (tim-bezhashvyly) [inline src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php]: `$productInOtherContexts` is never used.

File: src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php [hunk 3]
```diff
@@ -144,11 +173,11 @@ public function getStockQty()
     }
 
     /**
-     * @param string[] $contextDataSet
+     * @param Product[] $productInOtherContexts
      * @param string[] $otherValues
      * @return string
      */
-    public function getUrlKey(array $contextDataSet, array $otherValues)
+    public function getUrlKey(array $productInOtherContexts, array $otherValues)
```
Comment (tim-bezhashvyly) [inline src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php]: `$productInOtherContexts` is not used.

File: src/CatalogWriter/ProductXmlWriter.php [hunk 4]
```diff
@@ -19,14 +21,15 @@ public function setXmlWriter(\XMLWriter $xmlWriter)
...
     }
 
-    public function writeProduct(ProductSource $productSource)
+    public function writeProduct(Product $product)
     {
-        $this->startProduct($productSource);
-        $this->writeAttributes($productSource->getAttributeList());
+        $this->startProduct($product);
+        $this->writeImages($product->getImages());
+        $this->writeAttributes($product->getAttributes());
```
Comment (tim-bezhashvyly) [inline src/CatalogWriter/ProductXmlWriter.php]: Hm. Does the `getAttributes` method exist?

File: src/CatalogWriter/ProductXmlWriter.php [hunk 5]
```diff
@@ -100,4 +93,25 @@ private function endAttribute()
...
+    {
+        array_map(function (ProductImage $image) {
+            $this->writeImage($image);
+        }, $images->getImages());
+    }
+
+    private function writeImage(ProductImage $image)
+    {
+        $this->xmlWriter->startElement('image');
+        $this->writeAttributesContent($image->getAttributes());
```
Comment (tim-bezhashvyly) [inline src/CatalogWriter/ProductXmlWriter.php]: Same here.

**Comment by tim-bezhashvyly** [comment]
Review is done.

File: src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php (hunk 2 above)
Comment (Vinai) [inline src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php]: Yes, once I have time I would like to link some of the product values in different contexts, e.g. the categories and the images.

File: src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php (hunk 1 above)
Comment (Vinai) [inline src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php]: Fixed

File: src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php (hunk 3 above)
Comment (Vinai) [inline src/CatalogGenerator/ValueGenerator/SimpleProductValueGenerator.php]: See above... I don't have time to do it now :(

File: src/CatalogWriter/ProductXmlWriter.php (hunk 4 above)
Comment (Vinai) [inline src/CatalogWriter/ProductXmlWriter.php]: Yes, merged today: https://github.com/lizards-and-pumpkins/catalog/blob/master/src/Product/Product.php#L44

File: src/CatalogWriter/ProductXmlWriter.php (hunk 5 above)
Comment (Vinai) [inline src/CatalogWriter/ProductXmlWriter.php]: https://github.com/lizards-and-pumpkins/catalog/blob/master/src/Product/ProductImage.php#L60

**Pull request merged**